[Doc] Compatibility matrix for mutual exclusive features

wallashss opened the pull request:
Greetings,

We did a study of mutual exclusive features on vLLM and consolidated in a compatibility matrix.

We propose to add the compatibility matrix to the documentation pages to help users to quick consult to plan their implementation or study.

Following the table in markdown for quick checking and help reviewers. 

CC @njhill  @maxdebayser 

| | | | | | | | | | | | |
|-|-|-|-|-|-|-|-|-|-|-|-|
|Unnamed: 0|Chunked Prefill|APC|LoRa|Prompt Adapter|Speculative decoding|CUDA Graphs|Encoder/Decoder|Logprobs|Prompt Logprobs|Async Output|Multi-step|
|APC|✅|||||||||||
|LoRa|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L1558)]|✅||||||||||
|Prompt Adapter|✅|✅|✅|||||||||
|Speculative decoding|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L1200)] [[T](https://github.com/vllm-project/vllm/issues/5016)]|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/spec_decode/spec_decode_worker.py#L86-L87)]|✅||||||||
|CUDA Graphs|✅|✅|✅|✅|✅|||||||
|Encoder/Decoder|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L25)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L17)][[T](https://github.com/vllm-project/vllm/issues/7366)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L35)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L55)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L47)][[T](https://github.com/vllm-project/vllm/issues/7366)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L51)][[T](https://github.com/vllm-project/vllm/issues/7447)]||||||
|Logprobs|✅|✅|✅|✅|✅|✅|✅|||||
|Prompt Logprobs|✅|✅|✅|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/output_processor/multi_step.py#L65)] [[T](https://github.com/vllm-project/vllm/pull/8199)]|✅|✅|✅||||
|Async Output|✅|✅|✅|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L396)]|✅ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L383)]|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L200)] [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L51)]|✅|✅|||
|Multi-step|✗ [[C](https://github.com/vllm-project/vllm/blob/7de49aa86c7f169eb0962b6db29ad53fff519ffb/vllm/engine/arg_utils.py#L944)]|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/output_processor/multi_step.py#L130)]|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/arg_utils.py#L951)]|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L47)]|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/output_processor/multi_step.py#L65)][[T](https://github.com/vllm-project/vllm/issues/8198)]|✅||
|NVIDIA|✅|✅|✅|✅|✅|✅|✅|✅|✅|✅|✅|
|CPU|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L337)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L345)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L31)][[T](https://github.com/vllm-project/vllm/pull/4830)]|✗ [[T](https://github.com/vllm-project/vllm/issues/8475)]|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L327)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a84e598e2125960d3b4f716b78863f24ac562947/vllm/worker/cpu_model_runner.py#L125)]|✅|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L370)]|✗ [[T](https://github.com/vllm-project/vllm/issues/8477)]|
|AMD|✅|✅|✅|✅|✅|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/attention/backends/rocm_flash_attn.py#L343)]|✅|✅|✅|✗ [[T](https://github.com/vllm-project/vllm/issues/8472)]|

[C] = Link to where a check is made in the code and error reported
[T] = Link to open tracking issue or PR to address the incompatibility

DarkLight1337 commented:
This should also cut down on the number of issues flagged as bugs when in fact the feature is not supported yet.

Some comments:
- Is it intended that you omitted the first row (chunked prefill)? It kinda bothers me that the matrix is asymmetric.
- The table is getting quite long. I would split the table into two sections: one for core features and another one for hardware backends (by the way, you missed TPU backend).
- Let's also add a row for multimodal support under core features. cc @ywang96 
  - Chunked prefill: Not supported, WIP (#8346)
  - APC: Not supported, WIP (#8348)
  - LoRA: Not supported, WIP (#7199)
  - Speculative decoding: Not sure
  - CUDA Graphs: Supported
  - Encoder/decoder: Not supported (https://github.com/vllm-project/vllm/blob/main/vllm/inputs/preprocess.py#L300)
  - Logprobs: Supported
  - Prompt logprobs: Supported
  - Async output: Not sure
  - Multi-step: Not sure

wallashss commented:
Thanks for the feedback and the contribution on the multimodal feature @DarkLight1337 @ywang96 !

> Is it intended that you omitted the first row (chunked prefill)? It kinda bothers me that the matrix is asymmetric.

Yeah, kind of. The first time I added, the row become empty. Therefore I thought it would be waste to add it. I added again to you see it. If you think it looks nice I have no problem on keep it. 

> The table is getting quite long. I would split the table into two sections: one for core features and another one for hardware backends 

I discussed that before with a colleague, and we agreed to present like this in a concise format. What do you think? Do you think it looks like that bad? I can discuss this again and review it.

(by the way, you missed TPU backend).

Yeah, I know. There are some other devices that I did not add. Of course, there are some hanging fruits. But TPU and other devices I am currently not working with them and I don't have an easy setup to test like I did to the others. Would be nice if someone could contribute to add this information.

BTW, do you know or know someone that knows the support of multimodal on AMD and CPU? I assume that is supported on Nvidia by default. But do you know the minimum compute capability for this feature?

Here is the updated table:

<details>

<summary>TABLE</summary>

| | | | | | | | | | | | | |
|-|-|-|-|-|-|-|-|-|-|-|-|-|
|Feature|Chunked Prefill|APC|LoRa|Prompt Adapter|Speculative decoding|CUDA Graphs|Encoder/Decoder|Logprobs|Prompt Logprobs|Async Output|Multi-step|Multimodal|
|Chunked Prefill|||||||||||||
|APC|✅||||||||||||
|LoRa|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L1558)]|✅|||||||||||
|Prompt Adapter|✅|✅|✅||||||||||
|Speculative decoding|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L1200)] [[T](https://github.com/vllm-project/vllm/issues/5016)]|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/spec_decode/spec_decode_worker.py#L86-L87)]|✅|||||||||
|CUDA Graphs|✅|✅|✅|✅|✅||||||||
|Encoder/Decoder|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L25)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L17)][[T](https://github.com/vllm-project/vllm/issues/7366)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L35)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L55)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L47)][[T](https://github.com/vllm-project/vllm/issues/7366)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L51)][[T](https://github.com/vllm-project/vllm/issues/7447)]|||||||
|Logprobs|✅|✅|✅|✅|✅|✅|✅||||||
|Prompt Logprobs|✅|✅|✅|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/output_processor/multi_step.py#L65)] [[T](https://github.com/vllm-project/vllm/pull/8199)]|✅|✅|✅|||||
|Async Output|✅|✅|✅|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L396)]|✅ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L383)]|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L200)] [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L51)]|✅|✅||||
|Multi-step|✗ [[C](https://github.com/vllm-project/vllm/blob/7de49aa86c7f169eb0962b6db29ad53fff519ffb/vllm/engine/arg_utils.py#L944)]|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/output_processor/multi_step.py#L130)]|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/arg_utils.py#L951)]|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/worker/utils.py#L47)]|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/engine/output_processor/multi_step.py#L65)][[T](https://github.com/vllm-project/vllm/issues/8198)]|✅|||
|Multimodal|✗ [[T](https://github.com/vllm-project/vllm/pull/8346)]|✗ [[T](https://github.com/vllm-project/vllm/pull/8348)]|✗ [[T](https://github.com/vllm-project/vllm/pull/7199)]||?|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/main/vllm/inputs/preprocess.py#L300)]|✅|✅|?|||
|NVIDIA|✅|✅|✅|✅|✅|✅|✅|✅|✅|✅|✅|✅|
|CPU|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L337)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L345)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L31)][[T](https://github.com/vllm-project/vllm/pull/4830)]|✗ [[T](https://github.com/vllm-project/vllm/issues/8475)]|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/executor/cpu_executor.py#L327)]|✗  [[C](https://github.com/vllm-project/vllm/blob/a84e598e2125960d3b4f716b78863f24ac562947/vllm/worker/cpu_model_runner.py#L125)]|✅|✅|✗ [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/config.py#L370)]|✗ [[T](https://github.com/vllm-project/vllm/issues/8477)]|?|
|AMD|✅|✅|✅|✅|✅|✅|✗  [[C](https://github.com/vllm-project/vllm/blob/a2469127db6144eedb38d0b505287c0044e4ce06/vllm/attention/backends/rocm_flash_attn.py#L343)]|✅|✅|✅|✗ [[T](https://github.com/vllm-project/vllm/issues/8472)]|?|

</details>

DarkLight1337 commented:
> Yeah, kind of. The first time I added, the row become empty. Therefore I thought it would be waste to add it. I added again to you see it. If you think it looks nice I have no problem on keep it.

Looks better now, thanks.

> I discussed that before with a colleague, and we agreed to present like this in a concise format. What do you think? Do you think it looks like that bad? I can discuss this again and review it.

I'm not overly bothered by this so you can keep it as is if that's the plan.

> BTW, do you know or know someone that knows the support of multimodal on AMD and CPU? I assume that is supported on Nvidia by default. But do you know the minimum compute capability for this feature?

It's supported on both. This feature isn't tied to compute capability, it only depends on whether the model runner implements it.

DarkLight1337 commented on docs/source/serving/compatibility_matrix.rst:
```diff
@@ -0,0 +1,231 @@
+   * - Feature
+     - Chunked Prefill
+     - APC
+     - LoRa
```
```suggestion
     - LoRA
```

DarkLight1337 commented on docs/source/serving/compatibility_matrix.rst:
```diff
@@ -0,0 +1,231 @@
+     - ✗ `[T] <https://github.com/vllm-project/vllm/pull/8346>`__ 
+     - ✗ `[T] <https://github.com/vllm-project/vllm/pull/8348>`__ 
+     - ✗ `[T] <https://github.com/vllm-project/vllm/pull/7199>`__ 
+     - 
```
```suggestion
     - ?
```

njhill commented:
Thanks a lot for this @wallashss! We should get the first pass of this added to the docs imo while we're working on adding more rows/columns, and make more folks aware of it.

> > Is it intended that you omitted the first row (chunked prefill)? It kinda bothers me that the matrix is asymmetric.
> 
> Yeah, kind of. The first time I added, the row become empty. Therefore I thought it would be waste to add it. I added again to you see it. If you think it looks nice I have no problem on keep it.

Yeah it was my suggestion to keep the top right triangle empty so that we don't have a bunch of duplicate entries (which could also then become out of sync if we updated one and not its reflection). But it does make it harder to look at one row/column at a time and see all the features it works with.. you have to follow an L shape. Not really sure which is better.

> > I discussed that before with a colleague, and we agreed to present like this in a concise format. What do you think? Do you think it looks like that bad? I can discuss this again and review it.
> 
> I'm not overly bothered by this so you can keep it as is if that's the plan.

@wallashss assuming I'm the colleague, I'm fine with having it split to a separate table :), but it would still be good for the columns to be the same and lined up (and not sure whether that would be possible with markdown).

Some other features we might want to add:
- Beam search
- best-of-n generation
- Guided decoding

We could also shorten some of the other names to make things fit better (and can include a legend at the bottom if needed).. like "Spec decoding" or even "SD".

K-Mistele commented:
would it be possible to add a column for prefix caching? I have found it does not work on volta-arch CUDA devices (Nvidia v100 tesla) due to some triton operation being unsupported on the older device.

njhill commented:
@K-Mistele APC is (automatic) prefix caching :)

Agree expanding with specific compute capabilities would be good!

wallashss commented:
Thanks for the feedback @K-Mistele.

> I have found it does not work on volta-arch CUDA devices (Nvidia v100 tesla) due to some triton operation being unsupported on the older device.

How did you figure out that? Does vllm raise an exception/log warning with a clear message, or did it just run "fine" and you knew that was not right? The original idea was to split on Nvidia architecture, but in my research, at least for those features, it may not be necessary.

K-Mistele commented:
> Thanks for the feedback @K-Mistele.
> 
> > I have found it does not work on volta-arch CUDA devices (Nvidia v100 tesla) due to some triton operation being unsupported on the older device.
> 
> How did you figure out that? Does vllm raise an exception/log warning with a clear message, or did it just run "fine" and you knew that was not right? The original idea was to split on Nvidia architecture, but in my research, at least for those features, I thought it may not be necessary.

It raised an exception about an unsupported triton operation or something like that. I can reproduce it if you'd like to see the specific error although I don't think I have a screenshot of it right this second. I think this happens for chunked prefill too.

wallashss commented:
> It raised an exception about an unsupported triton operation or something like that. I can reproduce it if you'd like to see the specific error although I don't think I have a screenshot of it right this second. I think this happens for chunked prefill too.

Wow, if you could paste the errors here that would be really nice! Thanks.

K-Mistele commented:
> > It raised an exception about an unsupported triton operation or something like that. I can reproduce it if you'd like to see the specific error although I don't think I have a screenshot of it right this second. I think this happens for chunked prefill too.
> 
> Wow, if you could paste the errors here that would be really nice! Thanks.

Can do! I ran into the issue on a v100 Tesla 32GB, which is a volta device - not turing.

pooyadavoodi commented:
It may be good to remind PR authors about updating the compatibility matrix if relevant (e.g. using PR checklist). Just to make sure it remains up to date.

K-Mistele commented:
> > > It raised an exception about an unsupported triton operation or something like that. I can reproduce it if you'd like to see the specific error although I don't think I have a screenshot of it right this second. I think this happens for chunked prefill too.
> > 
> > 
> > Wow, if you could paste the errors here that would be really nice! Thanks.
> 
> Can do! I ran into the issue on a v100 Tesla 32GB, which is a volta device - not turing.

btw @wallashss I found open PRs with more details information on chunked prefill and APC issues on volta devices:

- For Chunked prefill on Nvidia Tesla v100 (Volta): #2729, #6723, #8024

wallashss commented:
THanks for that @K-Mistele 

This will be very useful, gonna update the table and propose a new one very soon.

K-Mistele commented:
> THanks for that @K-Mistele
> 
> This will be very useful, gonna update the table and propose a new one very soon.

no problem! Haven't gotten around to reproducing the APC issue but IIRC it was similar.

tjtanaa commented:
FYI as additional info to the compatibility matrix, as of commit cb3b2b9ba4a95c413a879e30e2b8674187519a93 . AMD Multi-Step Feature is working, however a combination of MultiStep + Chunked Prefill is only supported on CUDA.

Related issue: https://github.com/vllm-project/vllm/issues/9111#issuecomment-2395689152
Related PR: #9038

wallashss committed:
[Doc] Feature compatibility matrix

Signed-off-by: Wallas Santos <wallashss@ibm.com>

njhill commented:
@wallashss this can be added for LoRA + chunked prefill: https://github.com/vllm-project/vllm/pull/9057

Re @pooyadavoodi's comment above, perhaps you could include an addition to the pull request template in this same PR: https://github.com/vllm-project/vllm/blob/main/.github/PULL_REQUEST_TEMPLATE.md

wallashss commented:
Hey everyone,

Thank you for your contributions and feedback.

I did a major update on the matrix:

- Split on two matrix: feature x feature and feature x hardware. For the hardware matrix, I also split Nvidia on architectures due to the evidence of unsupported features on Volta (Thanks @K-Mistele for the help with that)
- Tried to shrink a lot to make them fit on the screen. I added some CSS to decrease the font in order to do that (I hope nobody bothers with that). I did some abbreviations, and to help identify those I added some links to the documentation of the feature or tooltips that appears when user do a mouse hover. 
- I removed the [C] and [T] labels, because I thought they were messing up the readability of the table. Therefore I only added the `✗` with a link for the issue (if it does have one).
- For the [C] labels (label for code check), I went to the source code that they pointed to and added a comment with a reminder to update the compatibility matrix if the combo becomes valid. I guess it solves the problem of updating the reference (permalink) and helps contributors/reviewers check in loco that there is something more to update.  What do you think @njhill?
 

<details>

<summary> Screenshots </summary>

![image](https://github.com/user-attachments/assets/2fe229fc-714c-4a14-96dc-303a641d8a90)

![image](https://github.com/user-attachments/assets/9376ba4e-23ba-4a9d-96f8-ac141fdc4a4c)



</details>

wallashss committed:
[Doc] Filled some cells for Multimodal models

Signed-off-by: Wallas Santos <wallashss@ibm.com>

njhill commented:
Thanks @wallashss for all of the hard work on this! Let's get it merged and we can make other adjustments as follow-ons.
